### PR TITLE
Optimize String.Replace(string, string)

### PR DESCRIPTION
**Pull request by stephentoub**
We can significantly improve its throughput for a few scenarios:
- If both oldValue and newValue are single characters in the form of strings, we can just delegate to the Replace(char, char) overload, which is much faster.
- If oldValue is a single character but newValue isn't, we can use IndexOf to find it, rather than an open-coded loop, making search times much faster for reasonably sized strings where the thing being searched for isn't immediate.
- If oldValue is multiple characters and the string being searched for isn't super frequent (e.g. doesn't repeat every few characters), we can significantly speed up throughput by using IndexOf to search for the whole string.  For example, replacing "\r\n" with "\n" in the contents of a typical file.

This does come at a measurable cost when the oldValue is really common and tightly packed, e.g. searching for "aa" in "aaaaaaaaaaaaaaaaaa", so we can decide whether the tradeoff is the right one.

Some example results:
- `"This is a test. This is only a test.".Replace("!", "")` is 4x faster.
- The source text for a 35-line .cs file containing the benchmarks I was running, with `src.Replace("\r\n", "\n")` is 2.3x faster.
- `guidWithDashes.Replace("-", "")` is 10% faster.
- **But** `"aaaaaaaaaaaaaaaaaaaaaa".Replace("a", "bbb")` and `"aaaaaaaaaaaaaaaaaaaaaa".Replace("aa", "bbb")` are both 2x slower, as we end up using IndexOf to search for `a` that's the very next character.

Basically it comes down to whether we believe vectorization is good or bad for this use case, whether the thing being searched for will be tightly packed or not.

Contributes to (or fixes?) https://github.com/dotnet/runtime/issues/44061

### CODE REVIEW

**Review by jkotas — APPROVED** [review]
LGTM. It is reasonable to assume that the improved cases are the common usage pattern.

**Review by GrabYourPitchforks — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs [hunk 1]
```diff
@@ -1124,48 +1124,71 @@ public string Replace(char oldChar, char newChar)
...
                 throw new ArgumentException(SR.Argument_StringZeroLength, nameof(oldValue));
+            }
 
-            // Api behavior: if newValue is null, instances of oldValue are to be removed.
-            newValue ??= string.Empty;
+            // If newValue is null, treat it as an empty string.  Callers use this to remove the oldValue.
+            newValue ??= Empty;
 
+            // Track the locations of oldValue to be replaced.
             var replacementIndices = new ValueListBuilder<int>(stackalloc int[StackallocIntBufferSizeLimit]);
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:1140]: always (stack)allocated even if we go `Replace(oldValue[0], newValue[0]);` fast path

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:1140]: Is that a problem?

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:1140]: However, it's probably still better than putting two `stackalloc int[]` in those two branches 🤔

Comment (GrabYourPitchforks) [reply src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:1140]: Since the assembly is marked `[SkipLocalsInit]` this shouldn't result in significant overhead, right? Would need to crack open the codegen to see if anything changed.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:1140]: > Since the assembly is marked [SkipLocalsInit] this shouldn't result in significant overhead, right?

Right.

**Comment by stephentoub** [comment]
Failures are https://github.com/dotnet/runtime/issues/44037
